File: server/src/api/plugins.rs
//! `/api/plugins` — port of `Api/Plugins.cs` (ADR 0033, S9 #131). Index is
//! auth-only; install/delete are admin-gated. plugin-host itself is
//! unchanged — this proxies to it exactly as .NET did.

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::routing::get;
use axum::{Json, Router};
use serde::Deserialize;

use crate::auth::Claims;
use crate::error::{AppError, AppResult};
use crate::state::AppState;
use crate::{plugins, sources};

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/index", get(index))
        .route("/install", axum::routing::post(install))
        .route("/{id}", axum::routing::delete(delete_plugin))
}

async fn index(
    _claims: Claims,
    State(state): State<AppState>,
) -> AppResult<Json<Vec<plugins::PluginIndexEntry>>> {
    let entries = plugins::fetch_index(&state.config.plugin_index_url, &state.http)
        .await
        .ok_or(AppError::BadGateway)?;
    Ok(Json(entries))
}

#[derive(Deserialize, Default)]
struct InstallBody {
    #[serde(default)]
    plugin_id: Option<String>,
}

async fn install(
    claims: Claims,
    State(state): State<AppState>,
    Json(body): Json<InstallBody>,
) -> AppResult<StatusCode> {
    claims.require_admin()?;

    let plugin_id = body.plugin_id.unwrap_or_default();
    if plugin_id.trim().is_empty() {
        return Err(AppError::BadRequest("plugin_id required".into()));
    }

    let entries = plugins::fetch_index(&state.config.plugin_index_url, &state.http)
        .await
        .ok_or(AppError::BadGateway)?;
    let entry = entries
        .into_iter()
        .find(|e| e.id == plugin_id)
        .ok_or(AppError::NotFound)?;

    let Some(download_url) = entry.download_url.filter(|u| !u.trim().is_empty()) else {
        return Err(AppError::UnprocessableEntity(String::new()));
    };

    let effective_url = format!(
        "{}/{}",
        state.config.plugin_host_url.trim_end_matches('/'),
        plugin_id
    );

    // Check duplicate before calling plugin-host (saves a round-trip on conflicts).
    if sources::exists_by_base_url(&state.db, &effective_url).await? {
        return Err(AppError::Conflict(String::new()));
    }

    let install_url = format!(
        "{}/plugins/install",
        state.config.plugin_host_url.trim_end_matches('/')
    );
    let install_ok = state
        .http
        .post(&install_url)
        .json(&serde_json::json!({ "url": download_url }))
        .send()
        .await
        .map(|r| r.status().is_success())
        .unwrap_or(false);
    if !install_ok {
        return Err(AppError::BadGateway);
    }

    sources::insert_community_source(
        &state.db,
        &uuid::Uuid::new_v4().to_string(),
        &plugin_id,
        &entry.name,
        &effective_url,
        &entry.content_types.join(","),
        entry.default_explicit,
    )
    .await?;

    Ok(StatusCode::CREATED)
}

async fn delete_plugin(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> AppResult<StatusCode> {
    claims.require_admin()?;

    let source = sources::find_by_source_key(&state.db, &id)
        .await?
        .ok_or(AppError::NotFound)?;
    if !source.is_community {
        return Err(AppError::Forbidden);
    }

    // Best-effort — don't fail the request if plugin-host is unreachable.
    let uninstall_url = format!(
        "{}/plugins/{id}",
        state.config.plugin_host_url.trim_end_matches('/')
    );
    let _ = state.http.delete(&uninstall_url).send().await;

    sources::delete(&state.db, &source.id).await?;
    Ok(StatusCode::NO_CONTENT)
}

File: server-dotnet/Api/Plugins.cs
using System.Security.Claims;
using System.Text.Json.Serialization;
using ArrghServer.Data;
using ArrghServer.Data.Models;
using Microsoft.EntityFrameworkCore;

namespace ArrghServer.Api;

public static class Plugins
{
    public static RouteGroupBuilder MapPluginsRoutes(this RouteGroupBuilder group)
    {
        group.MapGet("/index", GetIndex).RequireAuthorization().WithSummary("List all available plugins from the index");
        group.MapPost("/install", InstallPlugin).RequireAuthorization().WithSummary("Install a community plugin from URL");
        group.MapDelete("/{id}", DeletePlugin).RequireAuthorization().WithSummary("Remove a community plugin (bundled plugins cannot be removed)");
        return group;
    }

    // ── GET /api/plugins/index ────────────────────────────────────────────────

    static async Task<IResult> GetIndex(IConfiguration config, IHttpClientFactory http)
    {
        var url = config["PluginIndexUrl"] ?? "file:///app/plugin-index.json";
        var entries = await FetchIndexAsync(url, http);
        if (entries is null) return Results.StatusCode(502);
        return Results.Ok(entries);
    }

    // ── POST /api/plugins/install ─────────────────────────────────────────────

    static async Task<IResult> InstallPlugin(
        InstallBody body, ClaimsPrincipal principal,
        AppDbContext db, IConfiguration config, IHttpClientFactory httpFactory)
    {
        if (!Sources.IsAdmin(principal)) return Results.Forbid();

        var pluginIndexUrl = config["PluginIndexUrl"] ?? "file:///app/plugin-index.json";
        var pluginHostUrl = config["PluginHostUrl"] ?? "http://localhost:4000";

        var index = await FetchIndexAsync(pluginIndexUrl, httpFactory);
        if (index is null) return Results.StatusCode(502);

        var entry = index.FirstOrDefault(e => e.Id == body.PluginId);
        if (entry is null) return Results.NotFound();

        if (string.IsNullOrWhiteSpace(entry.DownloadUrl))
            return Results.UnprocessableEntity();

        var effectiveUrl = $"{pluginHostUrl.TrimEnd('/')}/{body.PluginId}";

        // Check duplicate before calling plugin-host (saves a round-trip on conflicts)
        var already = await db.ExternalSources.AnyAsync(s => s.BaseUrl == effectiveUrl);
        if (already) return Results.Conflict();

        // Tell plugin-host to download + load the bundle
        var http = httpFactory.CreateClient();
        var installUrl = $"{pluginHostUrl.TrimEnd('/')}/plugins/install";
        HttpResponseMessage installRes;
        try
        {
            installRes = await http.PostAsJsonAsync(installUrl, new { url = entry.DownloadUrl });
        }
        catch
        {
            return Results.StatusCode(502);
        }
        if (!installRes.IsSuccessStatusCode)
            return Results.StatusCode(502);

        db.ExternalSources.Add(new ExternalSource
        {
            Id = Guid.NewGuid().ToString(),
            Name = entry.Name,
            BaseUrl = effectiveUrl,
            ContentTypes = string.Join(",", entry.ContentTypes),
            Enabled = true,
            DefaultExplicit = entry.DefaultExplicit,
            IsCommunity = true,
            CreatedAt = DateTime.UtcNow,
        });
        await db.SaveChangesAsync();

        return Results.StatusCode(201);
    }

    // ── DELETE /api/plugins/{id} ──────────────────────────────────────────────

    static async Task<IResult> DeletePlugin(
        string id, ClaimsPrincipal principal,
        AppDbContext db, IConfiguration config, IHttpClientFactory httpFactory)
    {
        if (!Sources.IsAdmin(principal)) return Results.Forbid();

        var pluginHostUrl = config["PluginHostUrl"] ?? "http://localhost:4000";
        var effectiveUrl = $"{pluginHostUrl.TrimEnd('/')}/{id}";

        var source = await db.ExternalSources.FirstOrDefaultAsync(s => s.BaseUrl == effectiveUrl);
        if (source is null) return Results.NotFound();
        if (!source.IsCommunity) return Results.Forbid();

        // Tell plugin-host to unload — best-effort, don't fail on error
        try
        {
            var http = httpFactory.CreateClient();
            await http.DeleteAsync($"{pluginHostUrl.TrimEnd('/')}/plugins/{id}");
        }
        catch { }

        db.ExternalSources.Remove(source);
        await db.SaveChangesAsync();

        return Results.NoContent();
    }

    // ── Internal helpers ──────────────────────────────────────────────────────

    internal static async Task<List<PluginIndexEntry>?> FetchIndexAsync(string url, IHttpClientFactory httpFactory)
    {
        try
        {
            if (url.StartsWith("file://"))
            {
                var path = url["file://".Length..];
                var text = await File.ReadAllTextAsync(path);
                return System.Text.Json.JsonSerializer.Deserialize<List<PluginIndexEntry>>(text,
                    new System.Text.Json.JsonSerializerOptions { PropertyNameCaseInsensitive = true });
            }

            var http = httpFactory.CreateClient();
            return await http.GetFromJsonAsync<List<PluginIndexEntry>>(url,
                new System.Text.Json.JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }
        catch
        {
            return null;
        }
    }
}

// ── DTOs ──────────────────────────────────────────────────────────────────────

public class PluginIndexEntry
{
    [JsonPropertyName("id")]               public string Id { get; set; } = null!;
    [JsonPropertyName("name")]             public string Name { get; set; } = null!;
    [JsonPropertyName("description")]      public string? Description { get; set; }
    [JsonPropertyName("version")]          public string Version { get; set; } = null!;
    [JsonPropertyName("download_url")]     public string? DownloadUrl { get; set; }
    [JsonPropertyName("bundled")]          public bool? Bundled { get; set; }
    [JsonPropertyName("default_explicit")] public bool DefaultExplicit { get; set; }
    [JsonPropertyName("content_types")]    public List<string> ContentTypes { get; set; } = [];
}

public class InstallBody
{
    [JsonPropertyName("plugin_id")] public string PluginId { get; set; } = null!;
}
